R/install.R

```diff
@@ -1,276 +1,343 @@
+  )
 
-  bar_fmt <- if (isTRUE(progress)) {
-    collapse(sep = " | ", c(
-      "[:current/:total] :elapsedfull",
-      "ETA: :eta",
-      if (num_workers > 1) ":processes/:num_workers",
-      ":packages"))
-  } else {
-    ""
-  }
+  config <- list(lib = lib, num_workers = num_workers)
+  state <- make_start_state(plan, config)
 
-  bar <- progress::progress_bar$new(
-    total = if (!is.null(plan)) NROW(plan) else length(filenames),
-    format = bar_fmt,
-    stream = stdout(),
-    show_after = 0
-  )
+  withCallingHandlers({
 
-  if (is.null(plan)) {
-    res <- lapply(seq_along(filenames), function(idx) {
-      file <- filenames[idx]
-      meta <- metadata[[idx]]
-      format_message <- inplace(function(x) bar$message(format(x)))
-      installed_path <- with_handlers(
-        pkginstall_installed = format_message,
-        pkginstall_built = format_message,
-        pkginstall_begin = format_message,
-        error = exiting(function(e) {
-          name <- basename(e$package %||% e$path %||% file)
-          bar$message(glue("{red_cross()} Failed {name}"))
-          stop(e)
-        }), {
-          bar$tick(0, tokens = list(packages = get_pkg_name(file)))
-          install_package(file, lib = lib, metadata = meta,
-                          vignettes = vignettes)
-        })
-
-      bar$tick(1, tokens = list(packages = get_pkg_name(file)))
-      installed_path
-    })
-
-    names(res) <- filenames
-    return(structure(res, class = "installation_results", elapsed = Sys.time() - start))
-  }
+    ## Initialise one task for each worker
+    for (i in seq_len(state$config$num_workers)) {
+      ## TODO: update progress bar here
+      task <- select_next_task(state)
+      state <- start_task(state, task)
+    }
 
-  running <- character()
-  results <- list()
-  processes <- list()
-  update_progress <- function(count) {
-    bar$tick(count, tokens = list(packages = collapse(running, ", "), processes = length(processes), num_workers = num_workers))
-  }
-  update_progress(0)
-
-  # Currently assumes the order of installation is not important
-  events <- get_events(list(plan = plan, processes = list(), results = list()), num_workers, lib, bar)
-  while(running_processes(events)) {
-    for (output in get_output(events)) {
-      is_running <- grepl("^Building", strip_style(output$output))
-      running <- union(running, sub("^Building ", "", strip_style(output$output[is_running])))
-      finished <- output$output[!is_running]
-
-      if (length(finished) > 0 && nzchar(finished)) {
-
-        # remove installed from running
-        installed <- strip_style(finished)[grepl("^. Installed ", strip_style(finished))]
-        installed <- sub(". Installed ([^[:space:]]+).*", "\\1", installed)
-
-        running <- setdiff(running, installed)
-        bar$message(finished)
-        update_progress(length(installed))
-      }
-      if (!bar$finished && length(running)) {
-        update_progress(0)
-      }
+    while (1) {
```

Comment:
Maybe this should just be a `repeat { }` block?

Reply:
ok